### DISCUSSION TITLE
Sharp required for standalone mode

### DISCUSSION THREAD

**Discussion by vasanthi-d** [vercel/next.js#40125] [Help]
I am using next.js standalone mode and I am following [with-docker](https://github.com/vercel/next.js/blob/canary/examples/with-docker/Dockerfile) example  quoted on Next.js for containerizing the app. 
When run and access the app, I get below error 

`Error: 'sharp' is required to be installed in standalone mode for the image optimization to function correctly`

I have installed 'sharp' and it is part of my node_modules. What am I missing?

**Comment by martinIvovIv** [comment]
I am having a similar issue.
From the official documentation:
`If sharp is already installed but can't be resolved, set the NEXT_SHARP_PATH environment variable such as export NEXT_SHARP_PATH=/tmp/node_modules/sharp. Then, build your project with next build. Finally, restart the server with either next start for production mode or node server.js for standalone mode.`
[https://nextjs.org/docs/messages/sharp-missing-in-production](https://nextjs.org/docs/messages/sharp-missing-in-production)

Try this approach, if it doesn't work, see below.

But I am not sure about setting the env variable.
I found some relevant [StackOverflow threads](https://stackoverflow.com/questions/71426665/how-run-next-js-with-node-sharp-for-docker) and if that doesn't work try adding it in docker as:
`ENV NEXT_SHARP_PATH=/tmp/node_modules/sharp`

**Comment by asrudii** [comment]
I solved with added several command in dockerfile : 
1. add "RUN npm i sharp" in deps section
2. add "COPY --from=deps /root/.npm /root/.npm" in builder section before RUN npm run build
3. add "ENV NEXT_SHARP_PATH=/app/node_modules/sharp" in runner section

also if you use same image for deps, builder, and runner. I just add "RUN npm i sharp" in runner section to solved this.

**Reply by yashmalik23-weplay** [reply]
Hi @asrudii , can you share the full dockerfile, package.json and env file. I can't seem to fix this even after doing all the changes you mentioned.

**Reply by asrudii** [reply]
![image](https://github.com/vercel/next.js/assets/16185961/e7d44875-117d-40a8-8693-dc42d576c88b)

**Comment by rgembalik** [comment]
I, on the other end, simply had to add 
```
ENV NEXT_SHARP_PATH=/app/node_modules/sharp
```
in my `Dockerfile`. Not sure why the app would otherwise complain that there is no sharp installed since it was already in the `node_modules`.

**Reply by denkochev** [reply]
The only thing that worked for me.

**Reply by iasenovets** [reply]
solved for me

**Reply by rusakovic** [reply]
chm. what is missing?
![image](https://github.com/vercel/next.js/assets/17801144/d8d471ed-2f26-480e-ba7d-b463dcf71993)

**Comment by vvbbnn00** [comment]
Referring to issue: https://github.com/vercel/next.js/issues/59516

I have solved by downgrading `sharp` into `0.32.6`

**Reply by brnpimentel** [reply]
i'm with 0.32.6 version and getting the error too.

**Reply by alimoli** [reply]
I can confirm it and this doesn't need any env variable to work.
Thank you for that 🙏

**Reply by StevenMarrocco** [reply]
thank you! this worked for us as well

**Reply by Kikobeats** [reply]
It should be possible to use the latest sharp version, can you give a try? 🙏

**Reply by emredoganm** [reply]
Thank you @vvbbnn00 it does work for me 💃🏻

**Reply by Getitdan** [reply]
Can confirm it works. Add this to your Dockerfile: `RUN yarn add sharp@0.32.6 --ignore-engines`

**Comment by Pfed-prog** [comment]
ENV  NEXT_SHARP_PATH=/dspyt/node_modules/sharp

worked for me

![image](https://github.com/vercel/next.js/assets/66903336/f5657b0e-a439-4804-a4e3-602420f20dd0)

**Comment by Kikobeats** [comment]
If you are stuck on an old version of sharp (like 0.32.6), I have news for you: The Vercel team has fixed things that prevented newer versions of sharp from working.  

I just shipped a project using the latest sharp version (`sharp@0.33.2`) and it worked fine 🙂 .

**Reply by AmirHosseinKarimi** [reply]
I use `sharp@0.33.2` and it doesn't work.

**Reply by rvuyyuru2** [reply]
yes same for me as well

**Comment by styfle** [comment]
This should be fixed now on canary. You can try it out with `npm i next@canary`

**Reply by AmirHosseinKarimi** [reply]
When will it be applied to the stable version?

**Reply by styfle** [reply]
Its on stable in `next@14.1.4`

**Comment by ltbittner** [comment]
I can not get sharp 0.33.2 to work for the life of me. I'm on next@canary and I've tried every single trick I've found online, and no matter what it does not work. Quite frustrating

**Comment by abhithkba** [comment]
Doing these 2 steps helped me solve the issue :

1. Using sharp version `0.32.6`

2. Add `ENV NEXT_SHARP_PATH /app/node_modules/sharp` to the Dockerfile.

**Reply by ramyNiranjan** [reply]
This works for me as well

**Reply by yoonnokdoo** [reply]
Thanks!

**Reply by hcastillaq** [reply]
yes, it work

**Reply by longnguyendev** [reply]
@abhithkba  yes, it work thank a lot form VietNam with love

**Comment by xereda** [comment]
In which phase of the Dockerfile should I add the environment variable? Build or runner?

**Reply by abhithkba** [reply]
I did for Runner

**Reply by xereda** [reply]
I did in both. Same error:

``` 
   ▲ Next.js 14.0.4
   - Local:        http://b2f3e2e633c1:3000
   - Network:      http://172.17.0.2:3000

 ✓ Ready in 435ms
 ⨯ Error: 'sharp' is required to be installed in standalone mode for the image optimization to function correctly. Read more at: https://nextjs.org/docs/messages/sharp-missing-in-production

```

**Reply by abhithkba** [reply]
Try using either sharp versions 0.32.6 , 0.33.0 or 0.33.1 .
Cuz latest version tend to be ineffective here.

The fix for the problem is now available in v14.0.5-canary.60 of Next.js where you can use latest sharp package.

**Reply by xereda** [reply]
I adjusted it leaving the version fixed to 0.32.6. It worked! 👏🏻 🥳

**Reply by alexanderkhivrych** [reply]
did not work for me :(

**Reply by VladSam42** [reply]
which version of next works with sharp@0.32.6?

**Comment by CaptainN** [comment]
Sometimes you can run in to trouble from an annoying npm bug (if you are using npm, instead of really anything better, like yarn or pnpm, or bun - anything) - basically, if you have a partial of optional deps in your node_modules folder on like an M1 (which means you don't have binaries for other platforms like linux/x86 - this is normal, usually npm will only add the optional dep you need to node_modules), and then you do npm i - it'll actually remove the unused optional packages from your package-lock.json file.

Why does it do this? Because npm.

Anyway, the fix is easy - delete node_modules, and package-lock.json and reinstall with no node_modules, to avoid npm getting confused. Then it will put the missing optional packages in your package-lock.json file. Then all you have to do is remember to do this whenever you add or change anything in your packages... This has worked for me almost every time when I'm forced to use npm. But really, just use anything else other than npm...

**Comment by kzgzhn** [comment]
For Next 12 and sharp@0.33, I did it through
COPY --from=builder --chown=nextjs:nodejs /app/node_modules/@img /app/node_modules/@img

Before version 0.32, binary modules were downloaded from GitHub, but now they are downloaded from npm. However, the "standalone" mode does not see this dependency. Therefore, when building the Docker image, I manually copied this dependency. Perhaps this can be done through **outputFileTracingIncludes**, but in Next 12 there is no outputFileTracingIncludes yet.

I use the same [Dockerfile](https://github.com/vercel/next.js/blob/canary/examples/with-docker/Dockerfile) for building, so I don't see the point in reinstalling sharp; I just copy the dependency.

**Reply by bilgehan-biricik** [reply]
this solution worked for me!

**Reply by SC0d3r** [reply]
for Next 13 this doesnt work

**Reply by kzgzhn** [reply]
@SC0d3r in version 13 I encountered a problem with permissions for some files in the public folder, I did this::
```
# Production image, copy all the files and run next
FROM base AS runner
WORKDIR /app

ENV NODE_ENV production
# Uncomment the following line in case you want to disable telemetry during runtime.
# ENV NEXT_TELEMETRY_DISABLED 1

RUN addgroup --system --gid 1001 nodejs
RUN adduser --system --uid 1001 nextjs

#COPY --from=builder /app/public ./public <— removed

# Set the correct permission for prerender cache
RUN mkdir .next
RUN chown nextjs:nodejs .next

# Automatically leverage output traces to reduce image size
# https://nextjs.org/docs/advanced-features/output-file-tracing

# new line
COPY --from=builder --chown=nextjs:nodejs /app/public ./public
COPY --from=builder --chown=nextjs:nodejs /app/.next/standalone ./
COPY --from=builder --chown=nextjs:nodejs /app/.next/static ./.next/static
# new line
COPY --from=builder --chown=nextjs:nodejs /app/node_modules/@img /app/node_modules/@img

USER nextjs

EXPOSE 3000

ENV PORT 3000

# server.js is created by next build from the standalone output
# https://nextjs.org/docs/pages/api-reference/next-config-js/output
CMD HOSTNAME="0.0.0.0" node server.js
```

this works for me, but I don't know how safe it is to give owner to nextjs:nodejs. Why did the vercel team do this 🤷‍♂️


but I'm not sure where your problem is, maybe you need to go through the [documentation](https://nextjs.org/docs/pages/api-reference/next-config-js/output#automatically-copying-traced-files) again

**Comment by quochuydev** [comment]
It works for me

`node: 16`
`next: 13.4.13`

```
FROM node:16-alpine

WORKDIR /app
COPY .next/standalone /app
COPY .next/static /app/.next/static
COPY public /app/public

RUN yarn add -E sharp@0.32.6 --ignore-workspace-root-check
ENV NEXT_SHARP_PATH=/app/node_modules/sharp

CMD touch .env && \
echo "NODE_TLS_REJECT_UNAUTHORIZED=0" >> .env && \
mv .env /app/.env && \
HOSTNAME=0.0.0.0 PORT=3000 node /app/server.js
```

**Comment by Valdermeyder** [comment]
I faced the same issue for NextJS v14.2.21 when I run in standalone mode for the Docker image `node:22-bookworm-slim`. The solution was to [install cross platform dependencies for the `sharp` dependency](https://sharp.pixelplumbing.com/install#cross-platform).

I came to this after I modified the NextJS code in `node_modules/next/dist/server/image-optimizer.js` to log the actual error in place where empty `catch` was used.
Looking on the latest NextJS [image-optimizer.js](https://github.com/vercel/next.js/blob/04692855b77fb63debe3f3aa870233f9d9818998/packages/next/src/server/image-optimizer.ts#L51) code it seems the error handling is improved in NextJS v15
